### PR TITLE
TAJO-1485 Datum 'Char' returned only 1byte.

### PR DESCRIPTION
**Pull request by charsyam**
It is because of Tajo-Client's encoding and decoding problem.

### CODE REVIEW

File: tajo-client/src/main/java/org/apache/tajo/storage/RowStoreUtil.java [hunk 1]
```diff
@@ -197,7 +198,11 @@ private RowStoreEncoder(Schema schema) {
           case NULL_TYPE: nullFlags.set(i); break;
           case BOOLEAN: bb.put(tuple.get(i).asByte()); break;
           case BIT: bb.put(tuple.get(i).asByte()); break;
-          case CHAR: bb.put(tuple.get(i).asByte()); break;
+          case CHAR:
+						System.out.println("T1: " + tuple.get(i).size() + " " + tuple.get(i).asChars());
```
Comment (blrunner) [inline tajo-client/src/main/java/org/apache/tajo/storage/RowStoreUtil.java]: It seems to be a debug code. Could you delete this?

**Comment by blrunner** [comment]
Could you add some unit test cases to TestTupleUtil?

**Comment by charsyam** [comment]
@blrunner Thank you for your review.
I added unittest in TestTupleUtil and remove System.out debug code.

**Comment by blrunner** [comment]
@charsyam 

I found the updated patch and the unit test ran successfully. But I would like more better if you add one more unit test case to TestInsertQuery as follows.
```
create table test1 (col1 char(5));
insert into test1 select 'a';
insert into test1 select 'abc';
insert into test1 select 'abcde';
select * from test1;
col1
-------------------------------
a
abc
abcde
```

**Comment by charsyam** [comment]
@blrunner Thank you.
I'll add unittest asap.

**Comment by charsyam** [comment]
@blrunner I updated testcases and code :)

**Comment by blrunner** [comment]
Would you trigger Travis or Jenkins again after rebasing?

**Comment by charsyam** [comment]
@blrunner I rebased it, and it passed :)

**Comment by blrunner** [comment]
Thanks @charsyam 

I'll review this today. :)

**Comment by blrunner** [comment]
Sorry for my late review.
I failed to merge this as follows.
```
Auto-merging tajo-core/src/test/java/org/apache/tajo/engine/query/TestInsertQuery.java
CONFLICT (content): Merge conflict in tajo-core/src/test/java/org/apache/tajo/engine/query/TestInsertQuery.java
Automatic merge failed; fix conflicts and then commit the result.
```
Could you rebase this again?

**Comment by jihoonson** [comment]
Hi @charsyam, truly sorry for late review.
Your patch looks good, but misses the default length of the char type. As you know, the length can be ignored when defining a schema, and it is handled as the char of the 1 byte length.

**Comment by charsyam** [comment]
@jihoonson Thanks. 
but, I think it is already handled by 
https://github.com/apache/tajo/pull/531
Is there any mistakes I missed?

**Comment by jihoonson** [comment]
It seems that some cases are missed.
Here is a failure example.
```
default> create table test1 (col1 char);
OK
default> insert into test1 select 'a';
(1 rows, 0.23 sec, 319 B inserted)
default> select * from test1;
col1
-------------------------------
ERROR: java.lang.ArrayIndexOutOfBoundsException
java.sql.SQLException: java.lang.ArrayIndexOutOfBoundsException
	at org.apache.tajo.jdbc.TajoResultSetBase.next(TajoResultSetBase.java:786)
	at org.apache.tajo.cli.tsql.DefaultTajoCliOutputFormatter.printResult(DefaultTajoCliOutputFormatter.java:105)
	at org.apache.tajo.cli.tsql.TajoCli.localQueryCompleted(TajoCli.java:558)
	at org.apache.tajo.cli.tsql.TajoCli.executeQuery(TajoCli.java:532)
	at org.apache.tajo.cli.tsql.TajoCli.executeParsedResults(TajoCli.java:447)
	at org.apache.tajo.cli.tsql.TajoCli.runShell(TajoCli.java:419)
	at org.apache.tajo.cli.tsql.TajoCli.main(TajoCli.java:692)
Caused by: java.io.IOException: java.lang.ArrayIndexOutOfBoundsException
	at org.apache.tajo.jdbc.FetchResultSet.nextTuple(FetchResultSet.java:87)
	at org.apache.tajo.jdbc.TajoResultSetBase.next(TajoResultSetBase.java:780)
	... 6 more
Caused by: com.google.protobuf.ServiceException: java.lang.ArrayIndexOutOfBoundsException
	at org.apache.tajo.client.QueryClientImpl.fetchNextQueryResult(QueryClientImpl.java:370)
	at org.apache.tajo.client.TajoClientImpl.fetchNextQueryResult(TajoClientImpl.java:134)
	at org.apache.tajo.jdbc.FetchResultSet.nextTuple(FetchResultSet.java:69)
	... 7 more
```
This is the master log.
```
2015-05-08 18:02:04,216 ERROR org.apache.tajo.master.TajoMasterClientService:
java.lang.ArrayIndexOutOfBoundsException
        at java.lang.System.arraycopy(Native Method)
        at org.apache.tajo.storage.RowStoreUtil$RowStoreEncoder.toBytes(RowStoreUtil.java:203)
        at org.apache.tajo.master.exec.NonForwardQueryResultFileScanner.getNextRows(NonForwardQueryResultFileScanner.java:181)
        at org.apache.tajo.master.TajoMasterClientService$TajoMasterClientProtocolServiceHandler.getQueryResultData(TajoMasterClientService.java:551)
        at org.apache.tajo.ipc.TajoMasterClientProtocol$TajoMasterClientProtocolService$2.callBlockingMethod(TajoMasterClientProtocol.java:549)
        at org.apache.tajo.rpc.BlockingRpcServer$ServerHandler.channelRead0(BlockingRpcServer.java:100)
        at org.apache.tajo.rpc.BlockingRpcServer$ServerHandler.channelRead0(BlockingRpcServer.java:61)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:182)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
        at org.apache.tajo.rpc.MonitorServerHandler.channelRead(MonitorServerHandler.java:70)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:130)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by charsyam** [comment]
@jihoonson could you retry travis ci? I think it is broken with other issue not this :)

**Comment by jihoonson** [comment]
Thanks @charsyam. The above bug looks fixed.
Additionally, I have three comments as follows.
* Please follow our convention for indent.
* TestInsertQuery is a class to test Tajo's insert function rather than select result. I think that TestStorages looks a good place to test the character type for several types of storage.
* Please consider the following case. I think that ```ValueTooLongForTypeCharactersException``` should be thrown instead of ```java.lang.ArrayIndexOutOfBoundsException```.

```
default> \dfs -cat hdfs://localhost:7020/test
cat: `hdfs://localhost:7020/test': Is a directory
default> \dfs -cat hdfs://localhost:7020/test/test.csv
1|a|1
2|bb|2
3|ccc|3
4|dddd|4
5|eeeee|5

default> create external table test ( id int, name char(3), score int ) using text with ('text.delimiter'='|') location 'hdfs://localhost:7020/test';
OK

default> \d test

table name: default.test
table path: hdfs://localhost:7020/test
store type: text
number of rows: unknown
volume: 40 B
Options: 
	'text.delimiter'='|'

schema: 
id	INT4
name	CHAR(3)
score	INT4

default> select * from test;
id,  name,  score
-------------------------------
ERROR: java.lang.ArrayIndexOutOfBoundsException
java.sql.SQLException: java.lang.ArrayIndexOutOfBoundsException
	at org.apache.tajo.jdbc.TajoResultSetBase.next(TajoResultSetBase.java:786)
	at org.apache.tajo.cli.tsql.DefaultTajoCliOutputFormatter.printResult(DefaultTajoCliOutputFormatter.java:105)
	at org.apache.tajo.cli.tsql.TajoCli.localQueryCompleted(TajoCli.java:558)
	at org.apache.tajo.cli.tsql.TajoCli.executeQuery(TajoCli.java:532)
	at org.apache.tajo.cli.tsql.TajoCli.executeParsedResults(TajoCli.java:447)
	at org.apache.tajo.cli.tsql.TajoCli.runShell(TajoCli.java:419)
	at org.apache.tajo.cli.tsql.TajoCli.main(TajoCli.java:692)
Caused by: java.io.IOException: java.lang.ArrayIndexOutOfBoundsException
	at org.apache.tajo.jdbc.FetchResultSet.nextTuple(FetchResultSet.java:87)
	at org.apache.tajo.jdbc.TajoResultSetBase.next(TajoResultSetBase.java:780)
	... 6 more
Caused by: com.google.protobuf.ServiceException: java.lang.ArrayIndexOutOfBoundsException
	at org.apache.tajo.client.QueryClientImpl.fetchNextQueryResult(QueryClientImpl.java:370)
	at org.apache.tajo.client.TajoClientImpl.fetchNextQueryResult(TajoClientImpl.java:134)
	at org.apache.tajo.jdbc.FetchResultSet.nextTuple(FetchResultSet.java:69)
	... 7 more
```

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/apache/tajo/503)

**Comment by charsyam** [comment]
@jihoonson could you review new patch? I applied your review
I moved storage test into TestStorages.
but remained 2 tests into TestInsertQuery
and fix indent. Thank you.

**Comment by jihoonson** [comment]
+1 LGTM!
Thanks for your contribution.

**Comment by jihoonson** [comment]
I forgot closing this pr.
Would you please close? Thanks.

**Pull request closed without merging**